### PR TITLE
Add lang file updates on version change

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
This PR enables lang file updates when Skript detects a version change.
There are a few questions where I'm looking for input though:
- Should we allow addons to opt into this feature? Should it only be for Skript? My thought is that addon opt in would only be possible for addons that handle language files the same as Skript (I'm not sure any do tbh)
- Right now, it *always* updates and creates a backup of a lang file if a version change is detected. This is because Skript has no built-in capability of detecting differences in *values* of Config objects, only the actual keys themselves. Should we add a way for Skript to check if the values themselves have changed? It might result in *slightly* longer startup time when the version changes, but it might be better than just assuming it's changed.

---
**Target Minecraft Versions:** Any
**Requirements:** N/A
**Related Issues:** Needs merged before:
- https://github.com/SkriptLang/Skript/pull/4108

### CODE REVIEW

**Comment by AyhamAl-Ali** [comment]
Noice PR ⚡ 

1- I don't know if any addon is using such system yet so IMO we can decide this later if we have get more feedback on it.
2- That's good IMO, a backup is never a bad choice.

However, I would like this to have an option (in config.sk) so that users can choose to toggle auto update lang files in case they want to manually update with their custom lang files

**Comment by APickledWalrus** [comment]
> However, I would like this to have an option (in config.sk) so that users can choose to toggle auto update lang files in case they want to manually update with their custom lang files

That could be useful, however outdated lang files may result in console errors (as reported for Structure API, see https://discord.com/channels/988998880794402856/1012003185302450206/1013589722586689588 - testing guild).

This is why I opted for backups, as it ensures lang files will be up to date (but also preserves a user's work if they care to update it)

**Comment by AyhamAl-Ali** [comment]
The choice relies on whether it's preferred to

1- Override the user lang files with the new lang files and force the users to adapt their custom lang files every update manually
2- Force the users to update their lang files (after having a similar error of Structure API) and adapt their changes once in a while (when a lang file changes in some update)

Based on the answer the default value of `auto-update-lang-files` should be set

**Comment by TheLimeGlass** [comment]
Can this pull request also target the old aliases file? Issue stated here with error https://github.com/SkriptLang/Skript/issues/4861

**Comment by APickledWalrus** [comment]
> Can this pull request also target the materials.json? Issue(s) stated here with error #4861

See https://github.com/SkriptLang/Skript/pull/5105

**Comment by APickledWalrus** [comment]
> The choice relies on whether it's preferred to
> 
> 1- Override the user lang files with the new lang files and force the users to adapt their custom lang files every update manually 2- Force the users to update their lang files (after having a similar error of Structure API) and adapt their changes once in a while (when a lang file changes in some update)
> 
> Based on the answer the default value of `auto-update-lang-files` should be set

With the latest commit, it will only create backups if the lang file has been changed.
Unfortunately, I think we do need to force them to update. I don't think it's a good idea to add an opt out here because of the potential for errors (especially because the error message isn't always clear). We provide backups, so it wouldn't be the worse to convert over. We can also mention lang file changes in the update log :D

File: src/main/java/ch/njol/skript/Skript.java [hunk 1]
```diff
@@ -1218,12 +1218,13 @@ public static Collection<SkriptAddon> getAddons() {
 	 * @return A {@link SkriptAddon} representing Skript.
 	 */
 	public static SkriptAddon getAddonInstance() {
-		final SkriptAddon a = addon;
-		if (a == null)
-			return addon = new SkriptAddon(Skript.getInstance())
-					.setLanguageFileDirectory("lang");
-		else
-			return a;
+		SkriptAddon a = addon;
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/Skript.java:1221]: This is messy, why is it being done like this?

File: src/main/java/ch/njol/skript/localization/Language.java [hunk 2]
```diff
@@ -289,19 +290,41 @@ private static boolean load(SkriptAddon addon, String name) {
...
 	
-	private static HashMap<String, String> load(@Nullable InputStream in, String name) {
+	private static HashMap<String, String> load(@Nullable InputStream in, String name, boolean tryUpdate) {
 		if (in == null)
 			return new HashMap<>();
+
 		try {
-			return new Config(in, name + ".lang", false, false, ":").toMap(".");
+			Config langConfig = new Config(in, name + ".lang", false, false, ":");
+			in.close();
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/localization/Language.java:299]: Can we be using try-with-resources here? Or at least proper handling for this.

**Review by Moderocky — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]
Good. Maybe when we move on from Java 1.8 we can use try-with-resources.

File: src/main/java/ch/njol/skript/config/SectionNode.java [hunk 3]
```diff
@@ -446,45 +446,74 @@ HashMap<String, String> toMap(final String prefix, final String separator) {
...
+	/**
+	 * Updates the values of this SectionNode based on the values of another SectionNode.
+	 * @param other The other SectionNode.
+	 * @param countValues Whether or not values should be counted.
+	 * If false, values from the other SectionNode will be copied to this one.
+	 * @param excluded Keys to exclude from this update.
+	 * @return True if there are differences in the keys of this SectionNode and the other SectionNode.
+	 * If countValues is true, values are also included in this difference check.
+	 */
+	public boolean setValues(final SectionNode other, boolean countValues, final String... excluded) {
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/config/SectionNode.java:469]: Maybe a better param name would be checkValues

File: src/main/java/ch/njol/skript/config/SectionNode.java [hunk 4]
```diff
@@ -446,45 +446,74 @@ HashMap<String, String> toMap(final String prefix, final String separator) {
...
-		for (final Node n : this) {
-			if (CollectionUtils.containsIgnoreCase(excluded, n.key))
+		return setValues(other, false, excluded);
+	}
+
+	/**
+	 * Updates the values of this SectionNode based on the values of another SectionNode.
+	 * @param other The other SectionNode.
+	 * @param countValues Whether or not values should be counted.
+	 * If false, values from the other SectionNode will be copied to this one.
```
Comment (TPGamesNL) [inline src/main/java/ch/njol/skript/config/SectionNode.java:464]: I think this should be phrased differently. The main function of this method is to copy values over, so it should be said that if countValues is true, this is no longer the case. 

However, I'm also uncertain if the method `setValues` should have an argument to do exactly not what the name implies.

**Review by TPGamesNL — CHANGES_REQUESTED** [review]

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**